src/Grpc.AspNetCore.Web/Internal/Base64PipeReader.cs

```
@@ -0,0 +1,193 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Grpc.AspNetCore.Web.Internal
+{
+    /// <summary>
+    /// Reads and decodes base64 encoded bytes from the inner reader.
+    /// </summary>
+    internal class Base64PipeReader : PipeReader
+    {
+        private readonly PipeReader _inner;
+        private ReadOnlySequence<byte> _currentInnerBuffer;
+        private ReadOnlySequence<byte> _currentDecodedBuffer;
+
+        public Base64PipeReader(PipeReader inner)
+        {
+            _inner = inner;
+        }
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            var consumedPosition = ResolvePosition(consumed);
+
+            _inner.AdvanceTo(consumedPosition);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            var consumedPosition = ResolvePosition(consumed);
+            var examinedPosition = ResolvePosition(examined);
+
+            _inner.AdvanceTo(consumedPosition, examinedPosition);
+        }
+
+        private SequencePosition ResolvePosition(SequencePosition base64Position)
+        {
+            // TODO - is there a better way to get the index of a position in a sequence?
+            var length = _currentDecodedBuffer.Slice(0, base64Position).Length;
+
+            // IMPORTANT - this logic assumes that SequencePosition will be multiples of 3.
+            // GetMaxEncodedToUtf8Length will round up to next multiple of 3, so more data
+            // could be advanced internally than expected if not called using a mulitple of 3.
+            //
+            // Because ReadAsync always returns data in multiples of 3, and gRPC only uses
+            // ReadResult.Buffer.End to advance, the position should always be a multiple
+            // of 3. The only time it won't be is at the end of a message, in which case
+            // rounding up is the correct thing to do to consume padding.
```

Can this throw an InvalidOperationException or something to fail early if `length % 3 != 0`?

---

If length % 3 != 0 then it was because you're at the end of the message. You still want to advance 4 to consume padding.

---

This Base64PipeReader will be exposed to developers as `HttpContext.Request.BodyReader`, right? This would mean that developers can pass in arbitrary `examined` SequencePositions to PipeReader.AdvanceTo. In practice it makes sense that examined will always be `ReadResult.Buffer.End`, but we cannot know that for sure.

I still think we should throw if `length % 3 != 0` and we aren't at the end of the message.